### PR TITLE
Highgoal

### PR DESCRIPTION
**Pull request by oldgalileo**
Shooter code.

### CODE REVIEW

File: src/org/usfirst/frc4904/robot/commands/ShooterStart.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.usfirst.frc4904.robot.commands;
+
+
+import org.usfirst.frc4904.robot.RobotMap;
+import edu.wpi.first.wpilibj.command.Command;
+import edu.wpi.first.wpilibj.command.CommandGroup;
+
+public class ShooterStart extends CommandGroup {
+	public ShooterStart() {
+		addParallel(new FlywheelSpinup());
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/robot/commands/ShooterStart.java]: We should probably add the anti-jam somewhere here. Perhaps `addParallel` a `TimedCommand`?

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 2]
```diff
@@ -83,56 +88,63 @@
...
 	}
 
 	public RobotMap() {
 		Component.pdp = new PDP();
 		Component.shifter = new SolenoidShifters(Port.Pneumatics.solenoidUp, Port.Pneumatics.solenoidDown);
 		Component.navx = new NavX(SerialPort.Port.kMXP);
-		Component.chassisMC = new CustomPIDController(0.01, 0.0, -0.02, RobotMap.Component.navx);
-		Component.chassisMC.setInputRange(-180, 180);
-		Component.chassisMC.setContinuous(true);
+		Component.chassisTurnMC = new CustomPIDController(0.01, 0.0, -0.02, RobotMap.Component.navx);
```
Comment (billwpierce) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: Thanks for making these look nicer 👍

File: src/org/usfirst/frc4904/robot/commands/ShooterStart.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import edu.wpi.first.wpilibj.command.Command;
+import edu.wpi.first.wpilibj.command.CommandGroup;
+
+public class ShooterStart extends CommandGroup {
+	public final static double unloadDuration = 0.25; // TODO
+
+	public ShooterStart() {
+		addParallel(new FlywheelSpinup());
+		addParallel(new RunFor(new IndexerUnload(), ShooterStart.unloadDuration));
+		addSequential(new Command() {
```
Comment (leijurv) [inline src/org/usfirst/frc4904/robot/commands/ShooterStart.java]: Could use a subclass of command and a functional interface here... =)

**Review by c0d3rman — APPROVED** [review]
I took another look and can't find anything to improve, @HowardStark @billwpierce @leijurv please take another look before we test tomorrow if you get a chance

File: .classpath [hunk 4]
```diff
@@ -5,7 +5,7 @@
 	<classpathentry kind="var" path="networktables" sourcepath="networktables.sources"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/navx_frc.jar" sourcepath="USERLIBS_DIR/navx_frc-sources.jar"/>
-        <classpathentry kind="var" path="USERLIBS_DIR/CTRLib.jar"/>
+	<classpathentry kind="var" path="USERLIBS_DIR/CTRLib.jar"/>
```
Comment (oldgalileo) [inline .classpath]: lol

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 5]
```diff
@@ -45,15 +47,16 @@
 			public static final int ballioElevatorAndIntakeRoller = 3;
 			public static final int climbMotorA = 4;
 			public static final int climbMotorB = 5;
-			public static final int flywheelLeftMotor = 6; // WIP
-			public static final int flywheelRightMotor = 7; // WIP
+			public static final int flywheelMotorA = 6; // WIP
```
Comment (billwpierce) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: Why did you change them from flywheel left and right? Wouldn't left and right make more sense, and be easier for debugging when we're having a problem with one motor vs the other? Just a thought.

**Review by c0d3rman — APPROVED** [review]

File: src/org/usfirst/frc4904/robot/RobotMap.java (hunk 5 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/robot/RobotMap.java]: The two motors are treated identically in the code, so the distinction doesn't really make sense (just like the two motors for the left wheel). I would be OK with labelling them physically as A and B, though.

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 6]
```diff
@@ -138,7 +143,14 @@ public RobotMap() {
 			ballioDoorServo);
 		// Climber
 		Component.climber = new Climber(new CANTalon(Port.CANMotor.climbMotorA), new CANTalon(Port.CANMotor.climbMotorB));
-		Component.chassis = new TankDriveShifting("2017-Chassis", Component.leftWheel, Component.rightWheel, Component.shifter);
+		// Shooter
+		SpeedController flywheelMotorA = new CANTalon(Port.CANMotor.flywheelMotorA);
+		SpeedController flywheelMotorB = new CANTalon(Port.CANMotor.flywheelMotorB);
+		CustomEncoder flywheelEncoder = new CANEncoder("FlywheelEncoder", Port.CAN.flywheelEncoder);
```
Comment (carturn) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: The flywheel encoder will definitely be a CANTalonEncoder. I think this should be fixed before merging, even with a random CANTalon.

File: src/org/usfirst/frc4904/robot/commands/Shoot.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package org.usfirst.frc4904.robot.commands;
+
+
+import org.usfirst.frc4904.robot.RobotMap;
+import org.usfirst.frc4904.robot.subsystems.Shooter;
+import org.usfirst.frc4904.standard.commands.RunFor;
+import edu.wpi.first.wpilibj.command.Command;
+import edu.wpi.first.wpilibj.command.CommandGroup;
+
+public class Shoot extends CommandGroup {
```
Comment (carturn) [inline src/org/usfirst/frc4904/robot/commands/Shoot.java]: Is there a reason this command does not automatically bring the flywheel up to speed?

File: src/org/usfirst/frc4904/robot/commands/IndexerUnload.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package org.usfirst.frc4904.robot.commands;
+
+
+import org.usfirst.frc4904.robot.RobotMap;
+import org.usfirst.frc4904.robot.subsystems.Shooter;
+import org.usfirst.frc4904.standard.commands.motor.MotorConstant;
+
+public class IndexerUnload extends MotorConstant {
+	public IndexerUnload() {
+		super(RobotMap.Component.shooter.indexer, Shooter.INDEXER_UNLOAD_SPEED);
```
Comment (carturn) [inline src/org/usfirst/frc4904/robot/commands/IndexerUnload.java]: Why do we use RobotMap.Component.shooter.indexer here and the flywheel directly elsewhere?

**Review by carturn — CHANGES_REQUESTED** [review]
The shooter object is currently used somewhat inconsistently, although I think this is mostly an aesthetic issue.
The flywheel encoder should be changed from a CANEncoder to a CANTalonEncoder before merging.

File: src/org/usfirst/frc4904/robot/commands/Shoot.java (hunk 7 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/robot/commands/Shoot.java]: From an operator perspective, I want to be able to begin spinning up the flywheel while the driver is finishing alignment, so decoupling those actions lets me save cycle time. (To shoot automatically you can just run both Shoot and FlywheelSpinup in parallel.)

File: src/org/usfirst/frc4904/robot/commands/IndexerUnload.java (hunk 8 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/robot/commands/IndexerUnload.java]: Ah, good point - the flywheel should only be accessed through the shooter. The reason I made this mistake is because the flywheel is its own subsystem, while the indexer's just a motor.

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 9]
```diff
@@ -50,8 +53,11 @@
 			public static final int ballioDirectionalRoller = 1;
 			public static final int ballioHopperRollers = 2;
 			public static final int ballioElevatorAndIntakeRoller = 3;
-			public static final int flywheelLeftMotor = 6; // WIP
-			public static final int flywheelRightMotor = 7; // WIP
+			public static final int climbMotorA = 4;
+			public static final int climbMotorB = 5;
+			public static final int flywheelMotorA = 6; // WIP
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: Still WIP?

**Review by ajnadel — APPROVED** [review]

**Review by c0d3rman — APPROVED** [review]
Still good by me.

**Review by carturn — APPROVED** [review]
I am Carter, and I now approve this pull request.
